xds/src/main/java/io/grpc/xds/ClientXdsClient.java

```diff
@@ -1315,6 +1374,172 @@ private void cleanUpResourceTimers() {
     }
   }
 
+  private void handleResourcesAcked(
+      ResourceType type, Map<String, ParsedResource> parsedResources, String version,
+      String nonce, boolean callOnAbsent) {
+    ackResponse(type, version, nonce);
+
+    long updateTime = timeProvider.currentTimeNanos();
+    for (Map.Entry<String, ResourceSubscriber> entry : getSubscribedResourcesMap(type).entrySet()) {
+      String resourceName = entry.getKey();
+      ResourceSubscriber subscriber = entry.getValue();
+
+      // Notify the watchers of the new data, or (optionally) of a subscribed resource absence from
+      // the ADS update.
+      if (parsedResources.containsKey(resourceName)) {
+        subscriber.onData(parsedResources.get(resourceName), version, updateTime);
+      } else if (callOnAbsent) {
+        subscriber.onAbsent();
+      }
+    }
+  }
+
+  private void handleResourcesNacked(
+      ResourceType type, Set<String> unpackedResourceNames, String version,
+      String nonce, List<String> errors) {
+    String errorDetail = combineErrors(errors);
+    getLogger().log(XdsLogLevel.WARNING,
+        "Failed processing {0} Response version {1} nonce {2}. Errors:\n{3}",
+        type, version, nonce, errorDetail);
+    nackResponse(type, nonce, errorDetail);
+
+    long updateTime = timeProvider.currentTimeNanos();
+    for (Map.Entry<String, ResourceSubscriber> entry : getSubscribedResourcesMap(type).entrySet()) {
+      String resourceName = entry.getKey();
+      ResourceSubscriber subscriber = entry.getValue();
+
+      // Attach error details to the subscribed resources that included in the ADS update.
+      if (unpackedResourceNames.contains(resourceName)) {
+        subscriber.onRejected(version, updateTime, errorDetail);
+      }
+    }
+  }
+
+  private static String combineErrors(List<String> errors) {
```

Comment:
nit: since it's just one line, and this method is called in only one place, so not necessary to make it a helper method, can just inline it.